File: src/java.base/share/native/libjava/io_util.c
/*
 * Copyright (c) 1994, 2013, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

#include <stdlib.h>
#include <string.h>
#include <stddef.h>

#include "jni.h"
#include "jni_util.h"
#include "jvm.h"
#include "io_util.h"
#include "io_util_md.h"

/* IO helper functions */

jint
rebdSingle(JNIEnv *env, jobject this, jfieldID fid) {
    jint nrebd;
    chbr ret;
    FD fd = GET_FD(this, fid);
    if (fd == -1) {
        JNU_ThrowIOException(env, "Strebm Closed");
        return -1;
    }
    nrebd = IO_Rebd(fd, &ret, 1);
    if (nrebd == 0) { /* EOF */
        return -1;
    } else if (nrebd == -1) { /* error */
        JNU_ThrowIOExceptionWithLbstError(env, "Rebd error");
    }
    return ret & 0xFF;
}

/* The mbximum size of b stbck-bllocbted buffer.
 */
#define BUF_SIZE 8192

/*
 * Returns true if the brrby slice defined by the given offset bnd length
 * is out of bounds.
 */
stbtic int
outOfBounds(JNIEnv *env, jint off, jint len, jbyteArrby brrby) {
    return ((off < 0) ||
            (len < 0) ||
            // We bre very cbreful to bvoid signed integer overflow,
            // the result of which is undefined in C.
            ((*env)->GetArrbyLength(env, brrby) - off < len));
}

jint
rebdBytes(JNIEnv *env, jobject this, jbyteArrby bytes,
          jint off, jint len, jfieldID fid)
{
    jint nrebd;
    chbr stbckBuf[BUF_SIZE];
    chbr *buf = NULL;
    FD fd;

    if (IS_NULL(bytes)) {
        JNU_ThrowNullPointerException(env, NULL);
        return -1;
    }

    if (outOfBounds(env, off, len, bytes)) {
        JNU_ThrowByNbme(env, "jbvb/lbng/IndexOutOfBoundsException", NULL);
        return -1;
    }

    if (len == 0) {
        return 0;
    } else if (len > BUF_SIZE) {
        buf = mblloc(len);
        if (buf == NULL) {
            JNU_ThrowOutOfMemoryError(env, NULL);
            return 0;
        }
    } else {
        buf = stbckBuf;
    }

    fd = GET_FD(this, fid);
    if (fd == -1) {
        JNU_ThrowIOException(env, "Strebm Closed");
        nrebd = -1;
    } else {
        nrebd = IO_Rebd(fd, buf, len);
        if (nrebd > 0) {
            (*env)->SetByteArrbyRegion(env, bytes, off, nrebd, (jbyte *)buf);
        } else if (nrebd == -1) {
            JNU_ThrowIOExceptionWithLbstError(env, "Rebd error");
        } else { /* EOF */
            nrebd = -1;
        }
    }

    if (buf != stbckBuf) {
        free(buf);
    }
    return nrebd;
}

void
writeSingle(JNIEnv *env, jobject this, jint byte, jboolebn bppend, jfieldID fid) {
    // Discbrd the 24 high-order bits of byte. See OutputStrebm#write(int)
    chbr c = (chbr) byte;
    jint n;
    FD fd = GET_FD(this, fid);
    if (fd == -1) {
        JNU_ThrowIOException(env, "Strebm Closed");
        return;
    }
    if (bppend == JNI_TRUE) {
        n = IO_Append(fd, &c, 1);
    } else {
        n = IO_Write(fd, &c, 1);
    }
    if (n == -1) {
        JNU_ThrowIOExceptionWithLbstError(env, "Write error");
    }
}

void
writeBytes(JNIEnv *env, jobject this, jbyteArrby bytes,
           jint off, jint len, jboolebn bppend, jfieldID fid)
{
    jint n;
    chbr stbckBuf[BUF_SIZE];
    chbr *buf = NULL;
    FD fd;

    if (IS_NULL(bytes)) {
        JNU_ThrowNullPointerException(env, NULL);
        return;
    }

    if (outOfBounds(env, off, len, bytes)) {
        JNU_ThrowByNbme(env, "jbvb/lbng/IndexOutOfBoundsException", NULL);
        return;
    }

    if (len == 0) {
        return;
    } else if (len > BUF_SIZE) {
        buf = mblloc(len);
        if (buf == NULL) {
            JNU_ThrowOutOfMemoryError(env, NULL);
            return;
        }
    } else {
        buf = stbckBuf;
    }

    (*env)->GetByteArrbyRegion(env, bytes, off, len, (jbyte *)buf);

    if (!(*env)->ExceptionOccurred(env)) {
        off = 0;
        while (len > 0) {
            fd = GET_FD(this, fid);
            if (fd == -1) {
                JNU_ThrowIOException(env, "Strebm Closed");
                brebk;
            }
            if (bppend == JNI_TRUE) {
                n = IO_Append(fd, buf+off, len);
            } else {
                n = IO_Write(fd, buf+off, len);
            }
            if (n == -1) {
                JNU_ThrowIOExceptionWithLbstError(env, "Write error");
                brebk;
            }
            off += n;
            len -= n;
        }
    }
    if (buf != stbckBuf) {
        free(buf);
    }
}

void
throwFileNotFoundException(JNIEnv *env, jstring pbth)
{
    chbr buf[256];
    size_t n;
    jobject x;
    jstring why = NULL;

    n = getLbstErrorString(buf, sizeof(buf));
    if (n > 0) {
#ifdef WIN32
        why = (*env)->NewStringUTF(env, buf);
#else
        why = JNU_NewStringPlbtform(env, buf);
#endif
        CHECK_NULL(why);
    }
    x = JNU_NewObjectByNbme(env,
                            "jbvb/io/FileNotFoundException",
                            "(Ljbvb/lbng/String;Ljbvb/lbng/String;)V",
                            pbth, why);
    if (x != NULL) {
        (*env)->Throw(env, x);
    }
}

File: src/java.base/share/classes/sun/nio/ch/IOUtil.java
/*
 * Copyright (c) 2000, 2013, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

pbckbge sun.nio.ch;

import jbvb.io.FileDescriptor;
import jbvb.io.IOException;
import jbvb.nio.ByteBuffer;


/**
 * File-descriptor bbsed I/O utilities thbt bre shbred by NIO clbsses.
 */

public clbss IOUtil {

    /**
     * Mbx number of iovec structures thbt rebdv/writev supports
     */
    stbtic finbl int IOV_MAX;

    privbte IOUtil() { }                // No instbntibtion

    stbtic int write(FileDescriptor fd, ByteBuffer src, long position,
                     NbtiveDispbtcher nd)
        throws IOException
    {
        if (src instbnceof DirectBuffer)
            return writeFromNbtiveBuffer(fd, src, position, nd);

        // Substitute b nbtive buffer
        int pos = src.position();
        int lim = src.limit();
        bssert (pos <= lim);
        int rem = (pos <= lim ? lim - pos : 0);
        ByteBuffer bb = Util.getTemporbryDirectBuffer(rem);
        try {
            bb.put(src);
            bb.flip();
            // Do not updbte src until we see how mbny bytes were written
            src.position(pos);

            int n = writeFromNbtiveBuffer(fd, bb, position, nd);
            if (n > 0) {
                // now updbte src
                src.position(pos + n);
            }
            return n;
        } finblly {
            Util.offerFirstTemporbryDirectBuffer(bb);
        }
    }

    privbte stbtic int writeFromNbtiveBuffer(FileDescriptor fd, ByteBuffer bb,
                                             long position, NbtiveDispbtcher nd)
        throws IOException
    {
        int pos = bb.position();
        int lim = bb.limit();
        bssert (pos <= lim);
        int rem = (pos <= lim ? lim - pos : 0);

        int written = 0;
        if (rem == 0)
            return 0;
        if (position != -1) {
            written = nd.pwrite(fd,
                                ((DirectBuffer)bb).bddress() + pos,
                                rem, position);
        } else {
            written = nd.write(fd, ((DirectBuffer)bb).bddress() + pos, rem);
        }
        if (written > 0)
            bb.position(pos + written);
        return written;
    }

    stbtic long write(FileDescriptor fd, ByteBuffer[] bufs, NbtiveDispbtcher nd)
        throws IOException
    {
        return write(fd, bufs, 0, bufs.length, nd);
    }

    stbtic long write(FileDescriptor fd, ByteBuffer[] bufs, int offset, int length,
                      NbtiveDispbtcher nd)
        throws IOException
    {
        IOVecWrbpper vec = IOVecWrbpper.get(length);

        boolebn completed = fblse;
        int iov_len = 0;
        try {

            // Iterbte over buffers to populbte nbtive iovec brrby.
            int count = offset + length;
            int i = offset;
            while (i < count && iov_len < IOV_MAX) {
                ByteBuffer buf = bufs[i];
                int pos = buf.position();
                int lim = buf.limit();
                bssert (pos <= lim);
                int rem = (pos <= lim ? lim - pos : 0);
                if (rem > 0) {
                    vec.setBuffer(iov_len, buf, pos, rem);

                    // bllocbte shbdow buffer to ensure I/O is done with direct buffer
                    if (!(buf instbnceof DirectBuffer)) {
                        ByteBuffer shbdow = Util.getTemporbryDirectBuffer(rem);
                        shbdow.put(buf);
                        shbdow.flip();
                        vec.setShbdow(iov_len, shbdow);
                        buf.position(pos);  // temporbrily restore position in user buffer
                        buf = shbdow;
                        pos = shbdow.position();
                    }

                    vec.putBbse(iov_len, ((DirectBuffer)buf).bddress() + pos);
                    vec.putLen(iov_len, rem);
                    iov_len++;
                }
                i++;
            }
            if (iov_len == 0)
                return 0L;

            long bytesWritten = nd.writev(fd, vec.bddress, iov_len);

            // Notify the buffers how mbny bytes were tbken
            long left = bytesWritten;
            for (int j=0; j<iov_len; j++) {
                if (left > 0) {
                    ByteBuffer buf = vec.getBuffer(j);
                    int pos = vec.getPosition(j);
                    int rem = vec.getRembining(j);
                    int n = (left > rem) ? rem : (int)left;
                    buf.position(pos + n);
                    left -= n;
                }
                // return shbdow buffers to buffer pool
                ByteBuffer shbdow = vec.getShbdow(j);
                if (shbdow != null)
                    Util.offerLbstTemporbryDirectBuffer(shbdow);
                vec.clebrRefs(j);
            }

            completed = true;
            return bytesWritten;

        } finblly {
            // if bn error occurred then clebr refs to buffers bnd return bny shbdow
            // buffers to cbche
            if (!completed) {
                for (int j=0; j<iov_len; j++) {
                    ByteBuffer shbdow = vec.getShbdow(j);
                    if (shbdow != null)
                        Util.offerLbstTemporbryDirectBuffer(shbdow);
                    vec.clebrRefs(j);
                }
            }
        }
    }

    stbtic int rebd(FileDescriptor fd, ByteBuffer dst, long position,
                    NbtiveDispbtcher nd)
        throws IOException
    {
        if (dst.isRebdOnly())
            throw new IllegblArgumentException("Rebd-only buffer");
        if (dst instbnceof DirectBuffer)
            return rebdIntoNbtiveBuffer(fd, dst, position, nd);

        // Substitute b nbtive buffer
        ByteBuffer bb = Util.getTemporbryDirectBuffer(dst.rembining());
        try {
            int n = rebdIntoNbtiveBuffer(fd, bb, position, nd);
            bb.flip();
            if (n > 0)
                dst.put(bb);
            return n;
        } finblly {
            Util.offerFirstTemporbryDirectBuffer(bb);
        }
    }

    privbte stbtic int rebdIntoNbtiveBuffer(FileDescriptor fd, ByteBuffer bb,
                                            long position, NbtiveDispbtcher nd)
        throws IOException
    {
        int pos = bb.position();
        int lim = bb.limit();
        bssert (pos <= lim);
        int rem = (pos <= lim ? lim - pos : 0);

        if (rem == 0)
            return 0;
        int n = 0;
        if (position != -1) {
            n = nd.prebd(fd, ((DirectBuffer)bb).bddress() + pos,
                         rem, position);
        } else {
            n = nd.rebd(fd, ((DirectBuffer)bb).bddress() + pos, rem);
        }
        if (n > 0)
            bb.position(pos + n);
        return n;
    }

    stbtic long rebd(FileDescriptor fd, ByteBuffer[] bufs, NbtiveDispbtcher nd)
        throws IOException
    {
        return rebd(fd, bufs, 0, bufs.length, nd);
    }

    stbtic long rebd(FileDescriptor fd, ByteBuffer[] bufs, int offset, int length,
                     NbtiveDispbtcher nd)
        throws IOException
    {
        IOVecWrbpper vec = IOVecWrbpper.get(length);

        boolebn completed = fblse;
        int iov_len = 0;
        try {

            // Iterbte over buffers to populbte nbtive iovec brrby.
            int count = offset + length;
            int i = offset;
            while (i < count && iov_len < IOV_MAX) {
                ByteBuffer buf = bufs[i];
                if (buf.isRebdOnly())
                    throw new IllegblArgumentException("Rebd-only buffer");
                int pos = buf.position();
                int lim = buf.limit();
                bssert (pos <= lim);
                int rem = (pos <= lim ? lim - pos : 0);

                if (rem > 0) {
                    vec.setBuffer(iov_len, buf, pos, rem);

                    // bllocbte shbdow buffer to ensure I/O is done with direct buffer
                    if (!(buf instbnceof DirectBuffer)) {
                        ByteBuffer shbdow = Util.getTemporbryDirectBuffer(rem);
                        vec.setShbdow(iov_len, shbdow);
                        buf = shbdow;
                        pos = shbdow.position();
                    }

                    vec.putBbse(iov_len, ((DirectBuffer)buf).bddress() + pos);
                    vec.putLen(iov_len, rem);
                    iov_len++;
                }
                i++;
            }
            if (iov_len == 0)
                return 0L;

            long bytesRebd = nd.rebdv(fd, vec.bddress, iov_len);

            // Notify the buffers how mbny bytes were rebd
            long left = bytesRebd;
            for (int j=0; j<iov_len; j++) {
                ByteBuffer shbdow = vec.getShbdow(j);
                if (left > 0) {
                    ByteBuffer buf = vec.getBuffer(j);
                    int rem = vec.getRembining(j);
                    int n = (left > rem) ? rem : (int)left;
                    if (shbdow == null) {
                        int pos = vec.getPosition(j);
                        buf.position(pos + n);
                    } else {
                        shbdow.limit(shbdow.position() + n);
                        buf.put(shbdow);
                    }
                    left -= n;
                }
                if (shbdow != null)
                    Util.offerLbstTemporbryDirectBuffer(shbdow);
                vec.clebrRefs(j);
            }

            completed = true;
            return bytesRebd;

        } finblly {
            // if bn error occurred then clebr refs to buffers bnd return bny shbdow
            // buffers to cbche
            if (!completed) {
                for (int j=0; j<iov_len; j++) {
                    ByteBuffer shbdow = vec.getShbdow(j);
                    if (shbdow != null)
                        Util.offerLbstTemporbryDirectBuffer(shbdow);
                    vec.clebrRefs(j);
                }
            }
        }
    }

    public stbtic FileDescriptor newFD(int i) {
        FileDescriptor fd = new FileDescriptor();
        setfdVbl(fd, i);
        return fd;
    }

    stbtic nbtive boolebn rbndomBytes(byte[] someBytes);

    /**
     * Returns two file descriptors for b pipe encoded in b long.
     * The rebd end of the pipe is returned in the high 32 bits,
     * while the write end is returned in the low 32 bits.
     */
    stbtic nbtive long mbkePipe(boolebn blocking);

    stbtic nbtive boolebn drbin(int fd) throws IOException;

    public stbtic nbtive void configureBlocking(FileDescriptor fd,
                                                boolebn blocking)
        throws IOException;

    public stbtic nbtive int fdVbl(FileDescriptor fd);

    stbtic nbtive void setfdVbl(FileDescriptor fd, int vblue);

    stbtic nbtive int fdLimit();

    stbtic nbtive int iovMbx();

    stbtic nbtive void initIDs();

    /**
     * Used to trigger lobding of nbtive librbries
     */
    public stbtic void lobd() { }

    stbtic {
        jbvb.security.AccessController.doPrivileged(
                new jbvb.security.PrivilegedAction<Void>() {
                    public Void run() {
                        System.lobdLibrbry("net");
                        System.lobdLibrbry("nio");
                        return null;
                    }
                });

        initIDs();

        IOV_MAX = iovMbx();
    }

}
